### PR TITLE
Vanchor implement

### PR DESCRIPTION
**Pull request by filiplazovic**

### CODE REVIEW

File: contracts/vanchor/src/state.rs [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+
+        let new_root_index = (self.current_root_index + 1) % ROOT_HISTORY_SIZE;
+        self.current_root_index = new_root_index;
+        save_root(store, new_root_index, &current_level_hash)?;
+        self.next_index = next_index + 1;
+        Ok(next_index)
+    }
+
+    pub fn is_known_root(&self, root: [u8; 32], store: &dyn Storage) -> bool {
+        // if root == [0u8; 32] {
```
Comment (drewstone) [inline contracts/vanchor/src/state.rs:293]: Can we uncomment this?

File: contracts/vanchor/src/contract.rs [hunk 2]
```diff
@@ -0,0 +1,528 @@
...
+                element_encoder(&ArkworksIntoFieldBn254::into_field(calc_public_amt));
+            if calc_public_amt_bytes != proof_data.public_amount {
+                return Err(ContractError::Std(StdError::GenericErr {
+                    msg: "Invalid public amount".to_string(),
+                }));
+            }
+
+            // Construct public inputs
+            let chain_id_type_bytes = element_encoder(
+                &compute_chain_id_type(vanchor.chain_id, &COSMOS_CHAIN_TYPE).to_le_bytes(),
```
Comment (drewstone) [inline contracts/vanchor/src/contract.rs:276]: This should be BE bytes

File: contracts/vanchor/src/contract.rs [hunk 3]
```diff
@@ -0,0 +1,528 @@
...
+            for root in &proof_data.roots {
+                bytes.extend_from_slice(root);
+            }
+
+            let verifier = VANCHORVERIFIER.load(deps.storage)?;
+            let result = match (
+                proof_data.input_nullifiers.len(),
+                proof_data.output_commitments.len(),
+            ) {
+                (2, 2) => verify(verifier, bytes, proof_data.proof)?,
```
Comment (drewstone) [inline contracts/vanchor/src/contract.rs:298]: Can we add (16, 2) as well and task out integrating the 16-2 join/split tx. CC @dutterbutter to triage task assignment:
- Add 2 verifiers to the contract (small / large)
- Generate trusted parameters for 16-2 join/split (we should add ones for 2 vanchor bridge, 8 vanchor bridge, and 32 vanchor bridge)
- Add them to protocol-substrate-fixtures
- Hook them up here and add integration tests.

**Review by drewstone — CHANGES_REQUESTED** [review]
Looks good so far!

File: contracts/vanchor/src/state.rs (hunk 1 above)
Comment (duguorong009) [reply contracts/vanchor/src/state.rs:293]: The reason I commented this line is that the very first transact(deposit) is going with zero root.

File: contracts/vanchor/src/contract.rs (hunk 2 above)
Comment (duguorong009) [reply contracts/vanchor/src/contract.rs:276]: Hmm, I think it is "LE" bytes.
https://github.com/webb-tools/protocol-cosmwasm/blob/vanchor-implement/contracts/vanchor/src/test_util.rs#L201
This "test_util" is from "protocol-substrate" & it indicates that "LE" is used for chain id.

**Pull request merged**